File: src/php/cv.php
<?php

/**
 * PHP version 7
 * 
 * @category cv generator
 * @package  Template_Class
 * @author   Author <dmz.oneill@gmail.com>
 * @license  Apache2 license
 * @link     fio.ie
 * @file
 * Generate cv.php
 */

$cv_details = getenv("cv_file");
$template = getenv("template_file");
$out_html = getenv("output_file");

$template_html = file_get_contents($template);
$cv_yaml = file_get_contents($cv_details);
$cv_yaml = yaml_parse($cv_yaml);

/**
 * Re-usable function to replace strings or arrays of strings
 */
function section_replace_scalars($section, $replacements)
{
    if (stristr($section, "{list_item}")) {
        if (gettype($replacements) == "string") {
            $section = str_replace("{list_item}", $replacements . "\n", $section);
        } else {
            $details = "";
            foreach ($replacements as $X) {
                $copy = $section;
                $copy = str_replace("{list_item}", trim($X) . "\n", $copy);
                $details .= $copy;
            }
            $section = $details;
        }
    } else {
        preg_match_all("/\{(.*?)\}/", $section, $matches, PREG_OFFSET_CAPTURE);
        foreach ($matches[1] as $key => $value) {
            $section = str_replace("{" . $value[0] . "}", trim($replacements[$value[0]]) . "\n", $section);
        }
    }
    return $section;
}

/**
 * Replace references
 */
function references()
{
    global $template_html, $cv_yaml;
    preg_match("/<references>(.*?)<\/references>/ms", $template_html, $match);

    $replacement = "";
    foreach ($cv_yaml["references"] as $ref) {
        $replacement .= section_replace_scalars($match[1], $ref);
    }

    $template_html = preg_replace("/<references>(.*?)<\/references>/ms", $replacement, $template_html);
}

/**
 * Replace pages
 */
function pages()
{
    global $template_html, $cv_yaml;

    for ($count = 0; $count < count($cv_yaml["pages"]); $count++) {
        $X = strval($count + 1);
        preg_match("/<page" . $X . ">(.*?)<\/page" . $X . ">/ms", $template_html, $match);

        $page = $cv_yaml["pages"][$count];
        $jobs = $page["jobs"];

        $replacement = "";
        foreach ($jobs as $job) {
            $job_section = $match[1];

            preg_match("/<roles>(.*?)<\/roles>/ms", $job_section, $role_match);

            $roles_replacement = "";
            foreach ($job["roles"] as $role) {
                $role_section = $role_match[1];
                preg_match("/<details>(.*?)<\/details>/ms", $role_section, $details_match);

                $details_replacement = "";
                foreach ($role["details"] as $detail) {
                    $detail_copy = section_replace_scalars($details_match[1], $detail);
                    $details_replacement .= $detail_copy;
                }
                $role_section = preg_replace("/<details>(.*?)<\/details>/ms", $details_replacement, $role_section);
                $role_section = section_replace_scalars($role_section, $role);
                $roles_replacement .= $role_section;
            }

            $job_section = preg_replace("/<roles>(.*?)<\/roles>/ms", $roles_replacement, $job_section);
            $job_section = section_replace_scalars($job_section, $job);
            $replacement .= $job_section;
        }

        $template_html = preg_replace("/<page" . $X . ">(.*?)<\/page" . $X . ">/ms", $replacement, $template_html);
    }
}

/**
 * Replace roles
 */
function roles()
{
    global $template_html, $cv_yaml;

    $template_html = preg_replace("/{current-role}/ms", $cv_yaml["roles"][0], $template_html);

    preg_match("/<previous-roles>(.*?)<\/previous-roles>/ms", $template_html, $match);

    $rows = "";
    for ($count = 0; $count < count($cv_yaml["roles"]); $count++) {
        if ($count == 0) {
            continue;
        }

        $row = $match[1];
        $row = section_replace_scalars($row, $cv_yaml["roles"][$count]);
        $rows .= $row;
    }

    $template_html = preg_replace("/<previous-roles>(.*?)<\/previous-roles>/ms", $rows, $template_html);
}

/**
 * Replace remaining items
 */
function replace()
{
    global $template_html, $cv_yaml;

    references();
    pages();
    roles();

    foreach ($cv_yaml as $X => $value) {
        if (gettype($cv_yaml[$X]) == "string") {
            $replacement = $cv_yaml[$X];
            if (strstr($replacement, "\\n")) {
                $replacement = str_replace("\\n", "<br/>", $replacement);
            }
            $template_html = str_replace("{" . $X . "}", $cv_yaml[$X], $template_html);
        } else if (gettype($cv_yaml[$X]) == "array") {

            preg_match("/<" . $X . ">(.*?)<\/" . $X . ">/ms", $template_html, $match_instance);

            if (gettype($match_instance) != "array") {
                continue;
            }

            if (count($match_instance) == 0) {
                continue;
            }

            $list_replacement = "";
            foreach ($cv_yaml[$X] as $y) {
                $copy = $match_instance[1];
                if (gettype($y) == "array") {
                    if (isset($y["list"])) {
                        preg_match("/<children>(.*?)<\/children>/ms", $copy, $match);
                        $sublist = section_replace_scalars($match[1], $y["list"]);
                        $copy = preg_replace("/<children>(.*?)<\/children>/ms", $sublist, $copy);
                        $copy = section_replace_scalars($copy, $y["item"]);
                    } else {
                        $copy = preg_replace("/{list_item}.*?<ul.*?<\/ul>/ms", $y["item"], $copy);
                    }
                    $list_replacement .= $copy;
                } else if (gettype($y) == "string") {
                    $copy = section_replace_scalars($copy, $y);
                    $list_replacement .= $copy;
                }
            }

            $template_html = preg_replace("/<" . $X . ">(.*?)<\/" . $X . ">/ms", $list_replacement, $template_html);
        }
    }
}

replace();

file_put_contents($out_html, $template_html);

File: src/perl/cv.pl
#!/usr/bin/perl

use strict;
use warnings 'all';
use feature 'say';

use File::Slurp;
use YAML qw(LoadFile);
use Data::Dumper qw(Dumper);

my $cv_details = $ENV{'cv_file'};
my $template = $ENV{'template_file'};
my $out_html = $ENV{'output_file'};

my $template_html = read_file($template);
my %cv_yaml = %{ LoadFile($cv_details) };

sub section_replace_scalars
{
  my ($section, $replacements) = @_;

  print ">> section_replace_scalars\n";
  print $section . "\n";
  print $replacements . "\n";
  print ref($replacements) . "\n";

  if ($section =~ /{list_item}/) {
    print "{list_item}\n";
    if (ref($replacements) eq "") {
      print "string replacement\n";
      print $replacements . "\n";
      $section =~ s/{list_item}/$replacements/sm;
      print $section . "\n";
    } else {
      print "list replacement\n";
      my $details = "";
      foreach my $k (keys %{$replacements}) {
        my $copy = $section;
        $copy =~ s/{list_item}/%{$replacements}{$k}/sm;
        $details .= $copy;
      }
      $section = $details;
    }
  } else {
    my (@matches) = ($section =~ m/\{(.*?)\}/g);
    foreach my $match (@matches) {
      my $val = %$replacements{$match};
      $section =~ s/\{$match\}/$val/smg;
    }
  }

  return $section;
}

sub references
{
  my ($match) = ($template_html =~ m/<references>(.*?)<\/references>/ms);

  my $replacement = "";
  foreach my $ref (@{$cv_yaml{"references"}}) {
    $replacement .= section_replace_scalars($match, $ref);
  }

  $template_html =~ s/\<references>(.*?)<\/references>/$replacement/sm;
}


sub pages
{
  for (my $count = 1; $count <= 2; $count++) {
    my $match = $template_html =~ m/<page$count>(.*?)<\/page$count>/ms;
    my $page = $cv_yaml{"pages"}[$count];
    my $jobs = $page->{"jobs"};
    my $replacement = "";

    foreach my $job ($jobs) {
      my $job_section = $match;
      my $role_match = $job_section =~ m/<roles>(.*?)<\/roles>/ms;
      my $roles_replacement = "";

      foreach my $role ($job->{"roles"}) {
        my $role_section = $role_match;
        my $details_match = $role_section =~ m/<details>(.*?)<\/details>/ms;
        my $details_replacement = "";

        foreach my $detail ($role->{"details"}) {
          my $detail_copy = section_replace_scalars($details_match, $detail);
          $details_replacement .= $detail_copy;
        }

        $role_section =~ s/<details>.*?<\/details>/$details_replacement/smg;
        $role_section = section_replace_scalars($role_section, $role);
        $roles_replacement .= $role_section;
      }

      $job_section =~ s/<roles>.*?<\/roles>/$roles_replacement/smg;
      $job_section = section_replace_scalars($job_section, $job);
      $replacement .= $job_section;
    }

    $template_html =~ s/<page$count>.*?<\/page$count>/$replacement/smg;
  }
}

sub roles
{
  $template_html =~ s/{current-role}/%cv_yaml{"roles"}[0]/sm;

  if($template_html =~ m/<previous-roles>(.*?)<\/previous-roles>/ms){
    my $rows = "";
    my $count = 0;
    my $html = $1;
    foreach my $entry (%cv_yaml{"roles"}) {
      if ($count == 0) {
        $count = $count + 1;
        next;
      }

      my $row = $html;
      $row = section_replace_scalars($row, $entry);
      $rows .= $row;
    }

    $template_html =~ s/<previous-roles>(.*?)<\/previous-roles>/$rows/sm;
  }
}

sub replace
{
  references();
  # pages();
  roles();

  foreach my $X (keys %cv_yaml) {
    if (ref($cv_yaml{$X}) eq "") {
      my $replacement = $cv_yaml{$X};
      
      if(index($replacement, "\n") > 0) {
        $replacement =~ s/\n/<br\/>/sm;
      }

      $template_html =~ s/\{$X\}/$replacement/smg;
    } elsif (ref($cv_yaml{$X}) eq "ARRAY") {
      
      if($template_html =~ m/<$X>(.*?)<\/$X>/gms) {

        print "---\n";
        print $X . "\n";
        print $1 . "\n";
        # print $cv_yaml{$X} . "\n";
        # print ref($cv_yaml{$X}) . "\n";
        # print "+++\n";
        # print $#match_instance;
        # print @match_instance;
        # print "###\n";


        my $list_replacement = "";
        foreach my $y (%cv_yaml{$X}) {
          my $copy = $1;
          print "was \n";
          print $copy . "\n";
          if (ref($y) eq "ARRAY") {
            print @{$y}[0] . "\n";
            print "there \n";
            if (ref(@{$y}[0]) eq "HASH") {
              print "where \n";
              if($copy =~ m/<children>(.*?)<\/children>/ms) {
                print "here \n";
                print $1 . "\n";
                my $sublist = section_replace_scalars($1, @{$y}[0]);
                $copy =~ s/<children>(.*?)<\/children>/$sublist/sm;
                $copy = section_replace_scalars($copy, @{$y}[0]);
              }
            } else {
              $copy =~ s/{list_item}.*?<ul.*?<\/ul>/@{$y}[0]/sm;
            }
            $list_replacement .= $copy;
          } elsif (ref($y) eq "") {
            $copy = section_replace_scalars($copy, $y);
            $list_replacement .= $copy;
          }
        }

        $template_html =~ s/<$X>(.*?)<\/$X>/$list_replacement/sm;
      }
    }
  }
}

replace();

print $template_html;

open(my $fh, '>', $out_html);
print $fh $template_html;
close $fh;